agent/agent.go

```diff
@@ -0,0 +1,217 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+)
+
+type Agent struct {
+	cfg              Config
+	check            Checker
+	consulResultChan chan CheckResult
+	wsResultChan     chan CheckResult
+	webService       *webService
+	consul           *consul.Client
+	ctx              context.Context
+	ctxCancel        context.CancelFunc
+}
+
+type Config struct {
+	WebHost         string
+	WebPort         int
+	ServiceName     string
+	InstanceName    string
+	DefinitionsPath string
+	TTL             time.Duration
+}
+
+func New() (*Agent, error) {
+	config, err := DefaultConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create the agent configuration")
+	}
+
+	return NewWithConfig(config)
+}
+
+// returns a new instance of Agent with the given configuration
+func NewWithConfig(cfg Config) (*Agent, error) {
+	client, err := consul.NewClient(consul.DefaultConfig())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create a Consul client")
+	}
+
+	checker, err := NewChecker(cfg.DefinitionsPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create a Checker instance")
+	}
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+
+	wsResultChan := make(chan CheckResult, 1)
+
+	agent := &Agent{
+		cfg:          cfg,
+		check:        checker,
+		ctx:          ctx,
+		ctxCancel:    ctxCancel,
+		consul:       client,
+		webService:   newWebService(wsResultChan),
+		wsResultChan: wsResultChan,
+	}
+	return agent, nil
+}
+
+func DefaultConfig() (Config, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return Config{}, errors.Wrap(err, "could not read the hostname")
+	}
+
+	return Config{
+		InstanceName: hostname,
+		TTL:          10 * time.Second,
+	}, nil
+}
+
+func (a *Agent) Start() error {
+	log.Println("Registering the agent service with Consul...")
+	err := a.registerConsulService()
+	if err != nil {
+		return errors.Wrap(err, "could not register consul service")
+	}
+	log.Println("Consul service registered.")
+
+	defer func() {
+		log.Println("De-registering the agent service with Consul...")
+		err := a.consul.Agent().ServiceDeregister(a.cfg.InstanceName)
+		if err != nil {
+			log.Println("An error occurred while trying to deregisterConsulService the agent service with Consul:", err)
+		} else {
+			log.Println("Consul service de-registered.")
+		}
+	}()
+
+	var wg sync.WaitGroup
+	wg.Add(2)
```

Comment:
What is this `2` ?

Reply:
`2` is the number of items we want to wait for with the WaitGroup.
Here we start 2 goroutines we want to wait for, so we add `2` to the WG.
In each of those goroutines, below, `wg.Done()` is called: this does nothing more than decrementing the counter by one. Once the counter hits 0, the WG as a whole is considered done and `wg.Wait()` will finish blocking.

This whole mechanism is there to prevent goroutines from leaking (i.e. goroutines which are started but not ensured to actually stop before their caller terminates).